File: src/main/java/io/github/liquidTruffle/filters/BuiltinFilters.java
package io.github.liquidTruffle.filters;

import java.util.Map;

public class BuiltinFilters {
    public static void installInto(Map<String, Filter> out) {
        out.put("upcase", (input, args, kwargs) -> input != null ? input.toString().toUpperCase() : "");
        
        out.put("downcase", (input, args, kwargs) -> input != null ? input.toString().toLowerCase() : "");
        
        out.put("append", (input, args, kwargs) -> {
            Object rhs = args.isEmpty() ? "" : args.get(0);
            return (input != null ? input.toString() : "") + (rhs != null ? rhs.toString() : "");
        });
    }
}

File: src/main/kotlin/io/github/liquidTruffle/filters/BuiltinFilters.kt
package io.github.liquidTruffle.filters

object BuiltinFilters {
	fun installInto(out: MutableMap<String, Filter>) {
		out["upcase"] = object : Filter {
			override fun apply(input: Any?, args: List<Any?>, kwargs: Map<String, Any?>): Any? {
				return input?.toString()?.uppercase() ?: ""
			}
		}
		out["downcase"] = object : Filter {
			override fun apply(input: Any?, args: List<Any?>, kwargs: Map<String, Any?>): Any? {
				return input?.toString()?.lowercase() ?: ""
			}
		}
		out["append"] = object : Filter {
			override fun apply(input: Any?, args: List<Any?>, kwargs: Map<String, Any?>): Any? {
				val rhs = args.firstOrNull() ?: ""
				return "${input ?: ""}${rhs}"
			}
		}
	}
}
